torchci/lib/bot/greenBot.ts

```diff
@@ -0,0 +1,212 @@
+const LAND_ATTEMPT = "landed-attempt";
+const LAND_FAILED = "land-failed";
+
+type OctokitData = {
+  octokit: InstanceType<typeof ProbotOctokit>;
+  owner: string;
+  repo: string;
+  issue_number: number;
+  comment: string;
+};
+
+export default function greenBot(app: Probot): void {
+  app.on("issue_comment.created", async (ctx) => {
+    const commentBody = ctx.payload.comment.body;
+    if (commentBody.match(openOnGreen)) {
+      if (!ctx.payload.issue.pull_request) {
+        // Issue, not pull request.
+        await reactOnComment(ctx, "confused");
+      } else {
+        await reactOnComment(ctx, "+1");
+        await addLabels(ctx, [LAND_PENDING]);
+      }
+    }
+  });
+
+  app.on(["pull_request.labeled"], async (ctx) => {
+    if (ctx.payload.label.name === LAND_PENDING) {
+      // Check pull request's head ref and check runs to see if it has all green and merge if it's good
+      // TODO: Add pagination if checks go above 100
+      const payloadData = getPayloadData(ctx);
+      const ref = ctx.payload.pull_request.head.sha;
+      const prNum = ctx.payload.pull_request.number;
+      const checks = await ctx.octokit.checks.listForRef({
+        ...payloadData,
+        ref,
+        per_page: 100,
+      });
+
+      const [successfulChecks, failedChecks] = categorizeChecks(
+        checks.data.check_runs
+      );
+      // If all jobs are successful or neutral, then try merge, else do nothing
+      if (successfulChecks.length === checks.data.check_runs.length) {
+        await tryLandPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "All checks passed. Attempting to merge.",
+        });
+      }
+      // If any failed, alert the author
+      else if (failedChecks.length > 0) {
+        await failPR({
+          octokit: ctx.octokit,
+          ...payloadData,
+          issue_number: prNum,
+          comment: "Failed to land PR due to red signal.",
+        });
+      }
+    }
+  });
+
+  app.on("check_run.completed", async (ctx) => {
+    // Check that there are associated PRs that have the label
+    const prWithLabelsAwaitable = ctx.payload.check_run.pull_requests.map(
+      (pr) => {
+        return ctx.octokit.pulls.get({
+          ...getPayloadData(ctx),
+          pull_number: pr.number,
+        });
+      }
+    );
+    const prs = await Promise.all(prWithLabelsAwaitable);
+    const landPendingPrs = prs.filter((pr) => {
+      const ind = pr.data.labels.find((label) => {
+        return label.name === LAND_PENDING;
+      });
+      return ind != null;
+    });
+
+    if (landPendingPrs.length === 0) {
+      return;
+    }
+
+    if (landPendingPrs.length > 1) {
+      // If there's more than 1 land pending PR with the same checks,
+      // then we don't know which one to land
+      return;
+    }
+
+    const landPendingPr = prs[0];
+    // Check the other land signals
+    const checks = await ctx.octokit.checks.listForRef({
```

Comment:
This should be paginated, the full test suite exceeds 100 jobs. There is an octokit plugin for it.

Reply:
I think for checks, it's usually less than 100 (at least from what  I see on PRs) but yea would probably be good to paginate it.

Reply:
The default confit yes, but it goes over if someone specifies a ciflow tag